src/ORM/Hierarchy/Hierarchy.php

```diff
@@ -360,6 +381,127 @@ public function AllChildrenIncludingDeleted()
+     * Get the base class with the Hierarchy extension of a DataObject
+     * We need to get this instead of just getting the DataObjectSchema::baseDataClass() to ensure that we
+     * can run DB queries that filter on ParentID
+     */
+    public static function getBaseHierarchyClass(DataObject $obj): string
+    {
+        $baseHierarchyClass = $obj->ClassName;
+        $haveFoundTop = false;
+        do {
+            $parentDataClass = get_parent_class($baseHierarchyClass);
+            if (!$parentDataClass
+                || !$parentDataClass::has_extension(Hierarchy::class)
+                || !is_a($parentDataClass, DataObject::class, true)
+            ) {
+                $haveFoundTop = true;
+            } else {
+                $baseHierarchyClass = $parentDataClass;
+            }
+        } while (!$haveFoundTop);
+        return $baseHierarchyClass;
+    }
+
+    /**
+     * Inner recursive method used by getAllChildren
+     * This will use a hierarchical query method where it will initially fetched from the "second level down"
+     * where the ParentID equals the single DataObject at the "top level" of the hierarchy
+     * It will then fetch from the "third level down" in the hierarchy where the ParentID is of the ParentIDs in
+     * the "second level down". In will continue querying like this until the total number of cached objects has
+     * exceeded the configured value `node_threshold_total`, or all object have been fetched
+     */
+    private function recursivelyCacheDescendants(string $baseClass, int|array $idOrIDs): void
+    {
+        if (is_int($idOrIDs)) {
+            $parentIDs = [$idOrIDs];
+        } else {
+            $parentIDs = $idOrIDs;
+        }
+        $parentIDs = array_filter($parentIDs, function ($parentID) use ($baseClass) {
+            return !array_key_exists($parentID, Hierarchy::$all_children_child_ids_cache[$baseClass]);
+        });
+        if (empty($parentIDs)) {
+            return;
+        }
+        $config = $this->getOwner()->config();
+        $count = count(Hierarchy::$all_children_obj_cache[$baseClass]);
+        $threshold = $config->get('node_threshold_total');
+        if ($count >= $threshold) {
+            return;
+        }
+        $hideFromHierarchy = $config->get('hide_from_hierarchy');
+        $hideFromCMSTree = $config->get('hide_from_cms_tree');
+        $exclude = [];
+        if ($hideFromHierarchy) {
+            $exclude['ClassName'] = $hideFromHierarchy;
+        }
+        if ($hideFromCMSTree && $this->showingCMSTree()) {
+            $exclude['ClassName'] = $hideFromCMSTree;
+        }
+        $nextIDs = [];
+        /** @var DataList $children */
+        $children = $baseClass::get()->filter(['ParentID' => $parentIDs]);
+        if ($exclude) {
+            $children = $children->exclude($exclude);
+        }
+        foreach ($parentIDs as $parentID) {
+            Hierarchy::$all_children_child_ids_cache[$baseClass][$parentID] ??= [];
```

Comment:
```suggestion
            Hierarchy::$all_children_child_ids_cache[$baseClass][$parentID] = [];
```
We filtered by `!array_key_exists($parentID, Hierarchy::$all_children_child_ids_cache[$baseClass])` earlier, so the `??=` makes it look like we might have somehow set these inbetween that check and now which isn't possible from what I can tell.

Reply:
Updated